feat(tactic): command for adding tactic, command, and attribute documentation (RFC)

robertylewis opened the pull request:
As discussed in #1874, the current setup of `tactics.md`, etc isn't ideal. The entries often repeat information found in doc strings; they're located far from the definitions, so it's easy to forget to update them; the doc generator does something hackish to parse individual entries out of the markdown file.

This PR introduces a command `add_tactic_doc` that registers a "documentation entry" related to a declaration/list of declarations. It's best displayed by invoking it on itself:

````lean
add_tactic_doc
{ name := "add_tactic_doc",
  category := doc_category.cmd,
  decl_names := [`add_tactic_doc_command],
  tags := ["documentation"],
  description :=
"A command used to add documentation for a tactic, command, hole command, or attribute.

Usage: after defining an interactive tactic, command, or attribute, add its documentation as follows.
```
add_tactic_doc
{ name := \"display name of the tactic\",
  category := cat,
  decl_names := [`dcl_1, dcl_2],
  tags := [\"tag_1\", \"tag_2\"],
  description := Ädescribe what the command does here\"
}
```

The argument to `add_tactic_doc` is a structure of type `tactic_doc_entry`.
* `name` refers to the display name of the tactic; it is used as the header of the doc entry.
* `cat` refers to the category of doc entry.
  Options: `doc_category.tactic`, `doc_category.cmd`, `doc_category.hole_cmd`, `doc_category.attr`
* `decl_names` is a list of the declarations associated with this doc. For instance,
  the entry for `linarith` would set ``decl_names := [`tactic.interactive.linarith]``.
  Some entries may cover multiple declarations.
  It is only necessary to list the interactive versions of tactics.
* `tags` is an optional list of strings used to categorize entries.
* `description` is the body of the entry.
  What you are reading now is the description of `add_tactic_doc`.

If only one related declaration is listed in `decl_names` and it does not have a doc string,
`description` will be automatically added as its doc string.

Note that providing a badly formed `tactic_doc_entry` to the command can result in strange error messages.
"  }
````

Happy to hear any suggestions about this!

I haven't ported the entries from `tactics.md` yet. I'd rather get feedback first, merge this, and do that in a separate PR. Of course, the doc generation tool has to be updated too before deleting the .md files.

TO CONTRIBUTORS:

Make sure you have:

  * [ ] reviewed and applied the coding style: [coding](https://github.com/leanprover/mathlib/blob/master/docs/contribute/style.md), [naming](https://github.com/leanprover/mathlib/blob/master/docs/contribute/naming.md)
  * [ ] reviewed and applied [the documentation requirements](https://github.com/leanprover/mathlib/blob/master/docs/contribute/doc.md)
  * [ ] for tactics:
     * [ ] added or adapted documentation in [tactics.md](https://github.com/leanprover/mathlib/blob/master/docs/tactics.md)
     * [ ] write an example of use of the new feature in [tactics.lean](https://github.com/leanprover/mathlib/blob/master/test/tactics.lean)
  * [ ] make sure definitions and lemmas are put in the right files
  * [ ] make sure definitions and lemmas are not redundant

If this PR is related to a discussion on Zulip, please include a link in the discussion.

For reviewers: [code review check list](https://github.com/leanprover/mathlib/blob/master/docs/contribute/code-review.md)

robertylewis committed:
begin tactic_doc command

robertylewis committed:
merge add_tactic_doc with library_note

robertylewis committed:
fix library_note imports

robertylewis committed:
undo accidental revert

robertylewis committed:
better attribute description

robertylewis committed:
unneeded to_expr

robertylewis committed:
fix doc string attribution

bryangingechen commented on src/tactic/doc_commands.lean:
```diff
@@ -0,0 +1,186 @@
+  category := cat,
+  decl_names := [`dcl_1, dcl_2],
+  tags := [\"tag_1\", \"tag_2\"],
+  description := Ädescribe what the command does here\"
```
```suggestion
  description := \"describe what the command does here\"
```
Looks like something went wrong with the unicode input feature.

robertylewis committed:
unicode input error

Co-Authored-By: Bryan Gin-ge Chen <bryangingechen@gmail.com>

robertylewis committed:
display and collect doc entries

robertylewis committed:
missing doc string

gebner committed:
chore(*): switch to lean 3.6.0

robertylewis committed:
update for lean 3.6

robertylewis committed:
document core tactics

I considered adding these as doc strings in core. But the entry for `conv` mentions mathlib-specific things.

robertylewis committed:
move tfae and rcases docs

robertylewis committed:
move rintros and obtain

robertylewis committed:
simpa

robertylewis committed:
move part of tactic.interactive

robertylewis committed:
move more of tactic.interactive

robertylewis committed:
finish tactic.interactive

robertylewis committed:
move omega

robertylewis committed:
move renaming tactics

robertylewis committed:
move more

robertylewis committed:
move norm_cast

robertylewis committed:
more

robertylewis committed:
move more

bryangingechen committed:
doc commands in core + docstring tweaks

bryangingechen committed:
abel, alias, cache

bryangingechen committed:
elide, finish

bryangingechen committed:
hint, linearith, pi_instances

bryangingechen committed:
norm_num, rewrite

bryangingechen committed:
ring, ring_exp

bryangingechen committed:
solve_by_elim, suggest

bryangingechen committed:
doc_commands, rcases

bryangingechen committed:
ext

bryangingechen committed:
explode, find

bryangingechen committed:
lint

bryangingechen committed:
tidy

bryangingechen committed:
localized

bryangingechen committed:
reassoc_axiom, replacer, restate_axiom, where

bryangingechen committed:
simps

bryangingechen committed:
markdown files

bryangingechen committed:
interactive
